cpp/oneapi/dal/backend/primitives/reduction/reduction_rm_cw_dpc.cpp

```diff
@@ -0,0 +1,421 @@
+        : reduction_rm_cw_inplace(q, max_wg_size(q)) {}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+sycl::event reduction_rm_cw_inplace<Float, BinaryOp, UnaryOp>::operator()(
+    const Float* input,
+    Float* output,
+    const std::int64_t width,
+    const std::int64_t height,
+    const std::int64_t stride,
+    const BinaryOp& binary,
+    const UnaryOp& unary,
+    const event_vector& deps) const {
+    ONEDAL_ASSERT(0 < wg_ && wg_ <= max_wg_size(q_));
+    ONEDAL_ASSERT(0 <= width && width <= stride);
+    auto event = q_.submit([&](sycl::handler& h) {
+        h.depends_on(deps);
+        const auto range = get_range(width);
+        const auto kernel = get_kernel(input, output, height, stride, binary, unary);
+        h.parallel_for<kernel_t>(range, kernel);
+    });
+    return event;
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+sycl::event reduction_rm_cw_inplace<Float, BinaryOp, UnaryOp>::operator()(
+    const Float* input,
+    Float* output,
+    const std::int64_t width,
+    const std::int64_t height,
+    const BinaryOp& binary,
+    const UnaryOp& unary,
+    const event_vector& deps) const {
+    return this->operator()(input, output, width, height, width, binary, unary, deps);
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+sycl::nd_range<2> reduction_rm_cw_inplace<Float, BinaryOp, UnaryOp>::get_range(
+    const std::int64_t width) const {
+    return make_multiple_nd_range_2d({ width, wg_ }, { 1, wg_ });
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+typename reduction_rm_cw_inplace<Float, BinaryOp, UnaryOp>::kernel_t
+reduction_rm_cw_inplace<Float, BinaryOp, UnaryOp>::get_kernel(const Float* input,
+                                                              Float* output,
+                                                              const std::int64_t height,
+                                                              const std::int64_t stride,
+                                                              const BinaryOp& binary,
+                                                              const UnaryOp& unary) {
+    return kernel_t{ input,
+                     output,
+                     dal::detail::integral_cast<std::int64_t>(height),
+                     dal::detail::integral_cast<std::int32_t>(stride),
+                     binary,
+                     unary };
+}
+
+#define INSTANTIATE(F, B, U) template class reduction_rm_cw_inplace<F, B, U>;
+
+#define INSTANTIATE_FLOAT(B, U)                \
+    INSTANTIATE(double, B<double>, U<double>); \
+    INSTANTIATE(float, B<float>, U<float>);
+
+INSTANTIATE_FLOAT(min, identity)
+INSTANTIATE_FLOAT(min, abs)
+INSTANTIATE_FLOAT(min, square)
+
+INSTANTIATE_FLOAT(max, identity)
+INSTANTIATE_FLOAT(max, abs)
+INSTANTIATE_FLOAT(max, square)
+
+INSTANTIATE_FLOAT(sum, identity)
+INSTANTIATE_FLOAT(sum, abs)
+INSTANTIATE_FLOAT(sum, square)
```

Comment:
We usually instantiate all instances at the end of the file

Reply:
fixed